Parse <Globals> and <Placement>

RobertGemmaJr opened the pull request:
Uses functions to parse the `<Global>` and `PlacementRoot` children into an object.

`story` currently includes:
- `version` attribute
- `last_xpath` attribute
- `Camera`
- `CaveCamera`
- `background`
- `wand_navigation`
- `walls` (PlacementRoot)

closes #5 
closes #9 
closes #12 
closes #15

broarr commented on w3d_translator/read_xml.py:
```diff
@@ -39,12 +47,54 @@ def read_xml(file):
+    axis_xml = xml.find("Axis")
+    if axis_xml is not None:
+        axis = {
+            "rotation": str_to_tuple(axis_xml.attrib["rotation"]),
+            "angle": float(axis_xml.attrib["angle"]),
```
Totally agree here, looks like you could use it below too for the `look_at` dict

RobertGemmaJr replied:
Do you have any recommended reads on regex matching and that sort of thing in Python? Say for example running into a `(0.0, 0.0, -7.0)` type string (they're all over the xml) should always call `str_to_tuple`.

I did think/work through a recursive function to read the whole file at once but I found drilling through the objects for these more basic type conversions to be very tedious (the above would be something like `story["walls"]["Center"]["rotation"] = float(story["walls"]["Center]["rotation"])`.

RobertGemmaJr replied:
I'm also wondering if I should use asserts to validate the xml structure? No need to implement right now but given how unintuitive the xml already is (not to mention briefly reading through some of the other projects) I think it will pay off in the long run. Shall I add an issue?

broarr replied:
`import re; m = re.match("\(\s*-?(\d+(\.\d+)?),\s*-?(\d+(\.\d+)?),\s*-?(\d+(\.\d+)?)\)", "(0.0, 0.0, -7.0)")` should work some. The numbers are captured, but in groups 1, 3, and 5 (2, 4, and 6 are all the decimal portions of the numbers which are marked optional)

I'm into asserts! Throwing errors is a good thing

RobertGemmaJr replied:
I ended up checking for `bool`, `int`, `float`, and `tuple` using logic. Let me know if you think it can be tackled better/easier with regex though. I'm not super familiar with it so there'd be a learning curve for me first - definitely not opposed to learning

RobertGemmaJr replied:
Check that - switched `parse_string` to use regex. I'd appreciate a close set of eyes to make sure I set them up correctly 😄

broarr commented on w3d_translator/read_xml.py:
```diff
@@ -39,12 +47,51 @@ def read_xml(file):
+
+
+def parse_placement(xml: ET.Element) -> dict:
+    object = {
```
I'd change this name to `game_object` or something, just because the word `object` is way too overloaded. Not required, but might be nice

RobertGemmaJr replied:
I'll go with `placement` for now. I totally did just name it object as the defacto I just need a variable name here haha

broarr reviewed (commented):
This is looking great so far!

RobertGemmaJr commented on w3d_translator/read_xml.py:
```diff
@@ -1,13 +1,26 @@
+    story["background"] = str_to_tuple(g.find("Background").attrib["color"])
+    story["wand_navigation"] = parse_wand_navigation(g.find("WandNavigation"))
+
+    story["walls"] = parse_placements(root.find("PlacementRoot"))
```
Yeah you may be right. I want to differentiate the `<Placement>`'s in `<PlacementRoot` from those in the `Camera`, `Object`, etc tags. The "walls" are really reference points and will end up becoming GameObjects themselves in Unity whereas a Placement in the other roots are really just a relative position.

Perhaps I use `story["placements"]` instead of `story["walls"]` but rename the tag when I find it elsewhere? Struggling to come up with a pertinent variable name though

RobertGemmaJr commented on w3d_translator/read_xml.py:
```diff
@@ -39,12 +47,54 @@ def read_xml(file):
+        "position": position,
+        "Axis": axis,
+        "LookAt": look_at,
+    }
```
Done!

mcmcgrath13 commented on w3d_translator/read_xml.py:
```diff
@@ -50,51 +49,74 @@ def read_xml(file):
-        "LookAt": look_at,
-    }
+    # Check if string is a tuple (of integers or floats)
+    test = string.replace("(", "").replace(")", "").split(",")
```
a regex check seems probably a safer bet for this one

RobertGemmaJr replied:
Using regex now! Took a bit but I think I have it down

mcmcgrath13 commented on w3d_translator/read_xml.py:
```diff
@@ -50,51 +49,74 @@ def read_xml(file):
+
+def parse_attributes(xml: ET.Element) -> dict:
+    attributes = xml.attrib
+    for key, value in attributes.items():
+        attributes[key] = parse_string(value)
+    return attributes
```
in addition to having the parse_attributes, you could also have `parse_xml` which recurses on each of the children and `parse_xml`'s and then also parses attributes on the xml as well.

It still feels to me like you're explicitly walking the tree by finding things instead of just walking the tree and doing a rough pass at parsing which will get you to mostly the same place as far as I can tell

RobertGemmaJr replied:
There are definitely some things I want to change about the structure on the first pass. For example a `<Sound>` looks like

```xml
              <Sound name="gertrubience" filename="./audio/gertrubience.mp3" autostart="false">
                     <Mode>
                            <Fixed/> <!-- or <Positional /> -->
                     </Mode>
                     <Repeat>
                            <NoRepeat/> <!-- or <Repeat /> -->
                     </Repeat>
                     <Settings freq="1.0" volume="0.4" pan="0.0"/>
              </Sound>
```

I'd like to present this as one dict where `fixed` is a bool, `repeat` is a bool, and freq/volume/pan are indivdual keys (as opposed to a "Settings" key with the freq/volume/pan as a child dict). The children in the `<Global>` root aren't that bad but mostly everywhere else the xml has a really unintuitive structure. I don't want to recurse totally blindly - I'm going to have to walk the tree to lint this kind of stuff no matter what.

mcmcgrath13 replied:
I think it's fine to escape the default parsing path when needed, but I'd start with a generic parser and then add escape hatches

RobertGemmaJr replied:
So a recursive function that checks for specific tags?

mcmcgrath13 replied:
you know the code best - I'd just try to keep things generic as much as possible

RobertGemmaJr replied:
I did add a recursive function call but I'll be somewhat selective when using it

RobertGemmaJr commented:
Okay I refactored the PR a fair bit to make the parsing more general. Xml tags are TitleCase keys and attributes/leaves/text are snake_case.

- Added `parse_attributes` function (reads attributes from an XML tag)
- Added `parse_string` function (uses regex for string type conversion)
- Added `parse_recursive` function (recursively parses an xml tree)

eldu commented on w3d_translator/read_xml.py:
```diff
@@ -1,50 +1,97 @@
+
+        # Only add text property if it isn't empty
+        text = parse_string(xml.text) if xml.text is not None else None
+        if text != "" and text is not None:
+            val["text"] = text
```
I think this could be simplified to the following.
```suggestion
        if text:
            val["text"] = text
```

Both empty string and None are falsy

eldu reviewed (commented):
Some code clean up ideas. I'd suggest limiting the amount of regexs when you can since they are hard to read and most of the things that you're parsing for are things that can be accomplished through casting or testing inclusion. 

I would also assume that casting and testing inclusion are quicker than computing a regex, probably marginally and I don't have hard numbers for you on that one. They seem like simpler operations.

The tuples I think are a good case for a regex. Perhaps a comment on those. I tried looking at that one but was a bit confused at the format.

RobertGemmaJr commented on w3d_translator/read_xml.py:
```diff
@@ -1,50 +1,97 @@
-    for tag in root.find("ParticleActionRoot") or []:
-        pass  # Dict of ParticleActionList by name
+    # Check if string is a boolean
+    if re.match(r"^\s*(?i)(true)\s*$", string):
+        return True
+    elif re.match(r"^\s*(?i)(false)\s*$", string):
+        return False
```
`bool(string)` will always return true so I still need the if/else but will update!

eldu replied:
Oh true true! Since it's not an empty string! nice catch!

RobertGemmaJr commented on w3d_translator/read_xml.py:
```diff
@@ -1,50 +1,97 @@
-    story["background"] = Color(g.find("Background"))
-    story["wand_navigation"] = WandNavigation(g.find("WandNavigation"))
+    # Check if string is an integer
+    if re.match(r"^\s*-?(\d+)\s*$", string):
+        return int(string)
```
Done!

RobertGemmaJr commented on w3d_translator/read_xml.py:
```diff
@@ -1,50 +1,97 @@
 
-    return story
+    # Check if string is a float
+    if re.match(r"^\s*-?(\d+(\.\d+))\s*$", string):
+        return float(string)
```
Done!

RobertGemmaJr commented on w3d_translator/read_xml.py:
```diff
@@ -1,50 +1,97 @@
+
+    # Check if string is a tuple (of integers or floats)
+    if re.match(
+        r"\(?\s*-?(\d)\s*,\s*-?(\d)\s*,\s*-?(\d)?\)?",
```
They're actually groups of 3. 

Tuple of floats: `(3.2, -1.8, 2.6)` or `3.2, -1.8, 2.6`
Tuple of ints: `(3, -1, 2)` or `3, -1, 2`

eldu replied:
Oh i see that makes sense !  Thanks for adding in the comments :)

hollandjg commented:
I'd suggest adding some doctests as well, just so you're sure you're handling all the cases you think you are. I've put together a few for the parse_string – is it doing what you want in all the cases?

hollandjg reviewed (commented):
> > I've put together a few for the parse_string – is it doing what you want in all the cases?
> 
> @hollandjg where did you come up with the tests? I can definitely add a couple

Hi Robert! 
I just went through the code and put in what I thought it should be doing, given the comments in the code. (I didn't realize I had to submit the review so you can see it. Sorry about that!)

I wasn't sure though – it's not precisely clear what the intention is. Should more than 3 elements be handled in the tuple? What should happen if you get 4 elements? What's expected if you have a mixture of different elements? Should it loudly throw an exception, or should it just pass the string back as a string? Also, how does it relate to the recursive function in the same file? 

If you put in the cases you used yourself when generating the Regexes, then that's a great start.

RobertGemmaJr commented on w3d_translator/read_xml.py:
```diff
@@ -1,50 +1,105 @@
-    # sound_root = {}
-    for tag in root.find("SoundRoot") or []:
-        pass  # Dict of Sound by name
+def parse_string(string: str) -> Union[bool, int, float, tuple, Path, str]:
```
Ouu I really like that way of writing the tests I'll definitely add it!

hollandjg commented on w3d_translator/read_xml.py:
```diff
@@ -1,50 +1,182 @@
+    >>> parse_string('(13.5, 2.5, -1.5)')
+    (13.5, 2.5, -1.5)
+
+    # Test paths
```
How about this for the path testing?

```suggestion
    # Test Paths
    >>> p1 = parse_string('/folder/file.xml')
    >>> isinstance(p1, Path), p1.parts
    (True, ('/', 'folder', 'file.xml'))
    >>> p2 = parse_string('./folder/file.xml')
    >>> isinstance(p2, Path), p2.parts
    (True, ('folder', 'file.xml'))
    >>> p3 = parse_string('../folder/file.xml')
    >>> isinstance(p3, Path), p3.parts
    (True, ('..', 'folder', 'file.xml'))
    >>> p4 = parse_string('/folder/.')
    >>> isinstance(p4, Path), p4.parts
    (True, ('/', 'folder'))
    >>> p5 = parse_string('/folder/') # TODO: This is failing
    >>> isinstance(p5, Path), p5.parts
    (True, ('/', 'folder'))
    >>> p6 = parse_string('/folder')
    >>> isinstance(p6, Path), p6.parts
    (True, ('/', 'folder'))
    >>> p7 = parse_string('folder/file.xml') #TODO: 28
    >>> isinstance(p7, Path), p7.parts
    (True, ('folder', 'file.xml'))

```

hollandjg replied:
(This also deals with TODO 28. I think 28 looks like it's working at present.)

RobertGemmaJr commented:
I had some weird issue with `/` turning into `\\` and not being able to test for it so I just left the test as `isinstance(parse_string('/folder/file.xml'), Path)` and it's working! Thank you!
